### PR TITLE
[New Sample] Browse OGC API Feature Service

### PR DESCRIPTION
**Pull request by Rachael-E**
### Description

This PR is for the new OGC sample, Browse OGC API Feature Service. It is similar to our existing Browse WFS Sample, but with an updated UI and the use of an OGC API feature service rather than WFS. 

The PR also updates the Display OGC API Collection Sample to use the viewpoint changed listener, rather than the navigation changed listener following feedback from Colin. 

@jenmerritt could you take a first look please?

### CODE REVIEW

File: ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+import java.util.List;
+
+public class BrowseOgcApiFeatureServiceSample extends Application {
+
+  private ArcGISMap map;
+  private MapView mapView;
+  private ListView<OgcFeatureCollectionInfo> ogcLayerNamesListView;
+  private OgcFeatureService ogcFeatureService; // keep loadable in scope to avoid garbage collection
+
+  // UI class variables
```
Comment (jenmerritt) [inline ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java:59]: I quite like this, with the UI class variables being highlighted separately - maybe something to keep doing!

File: ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+    mapView = new MapView();
+    mapView.setMap(map);
+
+    // add the map view, UI controls, and the progress indicator to the stack pane
+    stackPane.getChildren().addAll(mapView, uiControlsVBox(), progressIndicator);
+
+    // when the load button is clicked, load the provided OGC API service url
+    loadButton.setOnAction(e -> loadOgcApiService(textField.getText()));
+
+    // when the load selected layer button is clicked, load the selected layer from the list view when the layer is selected
```
Comment (jenmerritt) [inline ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java]:
```suggestion
    // when the load selected layer button is clicked, load the selected layer from the list view
```

File: ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+   *
+   * @param serviceUrl the service url of the OGC feature service
+   */
+  private void loadOgcApiService(String serviceUrl) {
+
+    // clear any previously loaded OGC layers from the list view, and show the progress indicator
+    ogcLayerNamesListView.getItems().clear();
+    progressIndicator.setVisible(true);
+
+    // check if the text box is populated before instantiating a new OGC feature service from the provided URL, and loading it
```
Comment (jenmerritt) [inline ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java]:
```suggestion
    // check if the text box is populated before instantiating and loading a new OGC feature service from the provided URL
```

File: ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+    if (!textField.getText().isEmpty()) {
+
+      ogcFeatureService = new OgcFeatureService(serviceUrl);
+      ogcFeatureService.loadAsync();
+
+      // when the feature service has loaded, get its list of layers and add them to the list view
```
Comment (jenmerritt) [inline ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java]:
```suggestion
      // when the feature service has loaded, get a list of its layers and add them to the list view
```

File: ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+      });
+
+      // if the text field is blank and the load button is clicked, display an error
+    } else {
+      new Alert(Alert.AlertType.ERROR, "Enter a service url to continue").show();
+    }
+  }
+
+  /**
+   * Adds a OGCFeatureCollectionInfo to the map's operational layers.
```
Comment (jenmerritt) [inline ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java]:
```suggestion
   * Adds an OGCFeatureCollectionInfo to the map's operational layers.
```

File: ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+    map.getOperationalLayers().clear();
+
+    // create an OGC feature collection table from the feature collection info
+    var ogcFeatureCollectionTable = new OgcFeatureCollectionTable(ogcFeatureCollectionInfo);
+
+    // set the feature request mode to manual
+    // in this mode, the table must be manually populated as panning and zooming won't request features automatically.
+    ogcFeatureCollectionTable.setFeatureRequestMode(ServiceFeatureTable.FeatureRequestMode.MANUAL_CACHE);
+
+    // populate the table and then remove progress indicator and set the viewpoint to that of the layer's full extent when done
```
Comment (jenmerritt) [inline ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java]: Confused by this comment

File: ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+    var queryParameters = new QueryParameters();
+    // set a limit of 1000 on the number of returned features per request, the default on some services could be as low as 10
+    queryParameters.setMaxFeatures(1000);
+
+    try {
+      // populate the table with the query, leaving existing table entries intact, setting the outfields parameter to null requests all fields
```
Comment (jenmerritt) [inline ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java]:
```suggestion
      // populate the table with the query parameters
      // set the clearCache parameter to false to include existing table entries, and set the outfields parameter to null to request all fields
```
Just a suggestion, feel free to reword it just didn't quite read right.

File: ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java (hunk 6 above)
Comment (jenmerritt) [reply ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java]: Ah I see this maybe refers to the whole section below? I think maybe lose it, as it caused me confusion and I think you explain these on your line-by-line comments.

File: ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+      Envelope collectionExtent = ogcFeatureLayer.getFullExtent();
+      if (!collectionExtent.isEmpty()) {
+        mapView.setViewpointGeometryAsync(collectionExtent, 100);
+      }
+    });
+
+  }
+
+  /**
+   * Sets up the user interaction controls by adding a text field, load button, list view and load selected layer button in a VBox.
```
Comment (jenmerritt) [inline ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java]:
```suggestion
   * Sets up the user interaction controls by adding a text field, load button, list view and load selected layer button to a VBox.
```

File: ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+
+    // create a text field for entering a service URL, and provide an example service url to start with
+    textField = new TextField();
+    textField.setMaxWidth(300);
+    textField.setText("https://demo.ldproxy.net/daraa");
+
+    // create a button that allows the provided service to be loaded
+    loadButton = new Button("Load");
+
+    // create a list view to show all of the layers in a OGC API feature service
```
Comment (jenmerritt) [inline ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java]:
```suggestion
    // create a list view to show all of the layers in an OGC API feature service
```

File: ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+    ogcFeatureCollectionTable.setFeatureRequestMode(ServiceFeatureTable.FeatureRequestMode.MANUAL_CACHE);
+
+    // populate the table and then remove progress indicator and set the viewpoint to that of the layer's full extent when done
+    var queryParameters = new QueryParameters();
+    // set a limit of 1000 on the number of returned features per request, the default on some services could be as low as 10
+    queryParameters.setMaxFeatures(1000);
+
+    try {
+      // populate the table with the query, leaving existing table entries intact, setting the outfields parameter to null requests all fields
+      ListenableFuture<FeatureQueryResult> result = ogcFeatureCollectionTable.populateFromServiceAsync(queryParameters, false, null);
```
Comment (jenmerritt) [inline ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java:195]: So just to check this doesn't need the `ogcFeatureCollectionTable` to be loaded? As there next section has the `doneLoadingListener` and setups up the renderer etc. Making sure this doesn't need to be in there too / not sure at what point the table is loading.

**Review by jenmerritt — COMMENTED** [review]
Looks good @Rachael-E and certainly works well. Few comments around wording and one question just about at what point the `ogcFeatureCollectionTable` is loading and if anything else should be within the listener or not - probably will just benefit my understanding if anything!

File: ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java (hunk 6 above)
Comment (Rachael-E) [reply ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java]: I'm confused by this comment too! I suspect I accidentally left it from a previous iteration!

File: ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java (hunk 10 above)
Comment (Rachael-E) [reply ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java:195]: the loading of the ogcFeatureCollectionTable happens during the following code call, I've added a slight comment update to highlight this:

`ListenableFuture<FeatureQueryResult> result = ogcFeatureCollectionTable.populateFromServiceAsync(queryParameters, false, null);`

**Comment by Rachael-E** [comment]
thanks @jenmerritt ! I've addressed those comments, could you take another look please?

File: ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java (hunk 10 above)
Comment (jenmerritt) [reply ogc/browse-ogc-api-feature-service/src/main/java/com/esri/samples/browse_ogc_api_feature_service/BrowseOgcApiFeatureServiceSample.java:195]: Makes that part clear, nice!

**Review by jenmerritt — APPROVED** [review]
LGTM @Rachael-E

**Review by mbcoder — APPROVED** [review]
Happy with this sample

**Pull request merged**